doc/metadata-server/README.md

```diff
@@ -0,0 +1,327 @@
+
+- Marlowe contract metadata (`ContractMetadata`) - the same as in the previous section but now we can look at it as a product: `(InsMetadata, ExtMetadata)`.
+
+# Verification of `ContractMetadata`
+
+In the current scenario we want to verify the validity of two parts of `ContractMetadata` (`(InsMetadata, ExtMetadata)`). To do this we can slightly modify the previous verification schema. `InsMetadata` can be verified using the same strategy as before because it is information dedicated for a specific Contract so we need to prove the connection of the metadata with onchain triple: roles distribution, `Validator` and `Contract`.
+
+When we want to verify the `ExtContract` and `ExtMetadata` are really associated with the data on-chain we should prove that there is a set of parameters which when applied to `ExtContract` produced the same `Contract` value as we have in this particular `Tx`. `ExtContract` and `ExtMetadata` association should be proved by a separate signature / certificate.
+
+<!-- IMAGE Extended contract comparision flow. -->
+
+```purescript
+type RoleDistribution = Map Role PubKeyHash
+
+type Params = Map Id Marlowe.Semantics.Value
+
+type MarloweTxMetadata =
+  { instance ::
+    { -- | Here we have the same set of data as before
+    }
+  , template ::
+    { extContract :: ExtContract
+    , params :: Params
+    , metadata :: ExtMetadata
+    -- | Prove that `metadata` set is dedicated for the
+    -- | above `extContract` value.
+    , signature :: Signature
+    }
+  }
+
+type Metadata = { instance :: ..., template :: ... }
+
+extractMetadata :: OnChainTx -> Maybe Metadata
+extractMetadata tx = do
+  marloweTxMetadata <- decode <$> extractMetadata tx
+  txContract <- extractMarloweContract tx
+
+  let
+    templateTxMetadata = marloweTxMetadata.template
+
+  extMetadata <- do
+    if fillTemplate templateTxMetadata.extContract templateTxMetadata.params == txContract
+      && checkValidator txValidator
+      && checkSignature (templateTxMetadata.extContract /\ templateTxMetadata.metadata) MarloweTxMetadata.signature
+        then
+          Just templateTxMetadata.metadata
+        else
+          Nothing
+
+  -- | For the `InsMetadata` we reuse a similar procedure
+  -- | as in previous chapter
+  insMetadata <- extractInstanceMetadata
+
+  pure { insMetadata, extMetadata }
+```
+
+## Protocol
+
+### Objectives
+
+- We want to avoid centralization of metadata registry.
+
+- We want to minimize blockchain resource usage in the default setup.
+
+- We want to use [existing or emerging Cardano standards](https://cips.cardano.org/) and tools.
+
+- Metadata format schema should be well defined but we should leave it open / extensible.
```

Comment:
Could this be folded together with the ones in the motivation? Also they would be more powerful if they had justification